image/histogram_s4443048_Puyang_Deng/histogram.py

```diff
@@ -0,0 +1,158 @@
+import tensorflow as tf
+import warnings
+
+dtype_range = {tf.bool: (False, True),
+               tf.float16: (-1, 1),
+               tf.float32: (-1, 1),
+               tf.float64: (-1, 1)}
+
+def dtype_limits(image, clip_negative=False):
+    
+    """Return intensity limits, i.e. (min, max) tuple, of the image's dtype.
+    Parameters
+    ----------
+    image : ndarray
+        Input image.
+    clip_negative : bool, optional
+        If True, clip the negative range (i.e. return 0 for min intensity)
+        even if the image dtype allows negative values.
+    Returns
+    -------
+    imin, imax : tuple
+        Lower and upper intensity limits.
+    """
+    
+    imin, imax = dtype_range[image.dtype.type]
+    if clip_negative:
+        imin = 0
+    return imin, imax
+
+def _offset_array(arr, low_boundary, high_boundary):
+    
+    """Offset the array to get the lowest value at 0 if negative.
+    Parameters
+    ----------
+    arr : ndarray
+        Input image or array.
+    low_boundary : integer
+        The lower boundary intensity limit of the input image.
+    high_boundary : integer
+        The lower boundary intensity limit of the input image.
+    Returns
+    -------
+    arr, offset : tuple
+        Original input array and offset array to origin.
+    """
+    
+    if low_boundary < 0:
+        offset = low_boundary
+        arr = arr - offset
+    else:
+        offset = 0
+    return arr, offset
+
+def _bincount_histogram(image, source_range):
+    
+    """Efficient histogram calculation for an image of integers.
+    This function is significantly more efficient than self built histogram 
+    but works only on images of integers. It is based on self built bincount.
+    Parameters
+    ----------
+    image : array
+        Input image.
+    source_range : string
+        'image' determines the range from the input image.
+        'dtype' determines the range from the expected range of the images
+        of that data type.
+    Returns
+    -------
+    hist : array
+        The values of the histogram.
+    bin_centers : array
+        The values at the center of the bins.
+    """
+    
+    if source_range not in ['image']:
+        raise ValueError('Incorrect value for `source_range` argument: {}'.format(source_range))
+    
+    if source_range == 'image':
+        image_min = int(image.min().astype(tf.int64))
+        image_max = int(image.max().astype(tf.int64))
+        
+    elif source_range == 'dtype':
+        image_min, image_max = dtype_limits(image, clip_negative=False)
+        
+    image, offset = _offset_array(image, image_min, image_max)
```

Comment:
Good to add inline comments even if original code didn't have them

Reply:
Sorry for the delay sir! Just add the inline comment